src/components/tl/nccl/tl_nccl_team.c

```diff
@@ -0,0 +1,163 @@
+UCC_CLASS_DEFINE(ucc_tl_nccl_team_t, ucc_tl_team_t);
+
+ucc_status_t ucc_tl_nccl_team_destroy(ucc_base_team_t *tl_team)
+{
+    UCC_CLASS_DELETE_FUNC_NAME(ucc_tl_nccl_team_t)(tl_team);
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_nccl_team_create_test(ucc_base_team_t *tl_team)
+{
+    ucc_tl_nccl_team_t *team = ucc_derived_of(tl_team, ucc_tl_nccl_team_t);
+    ucc_status_t status;
+    ncclResult_t nccl_status;
+
+    status = team->oob.req_test(team->oob_req);
+    if (status == UCC_INPROGRESS) {
+        return UCC_INPROGRESS;
+    }
+    if (status != UCC_OK) {
+        team->oob.req_free(team->oob_req);
+        tl_error(tl_team->context->lib, "oob req test failed");
+        goto free_unique_id;
+    }
+    status = team->oob.req_free(team->oob_req);
+    if (status != UCC_OK) {
+        tl_error(tl_team->context->lib, "oob req free failed");
+        goto free_unique_id;
+    }
+    CUDACHECK_GOTO(cudaStreamCreateWithFlags(&team->stream,
+                   cudaStreamNonBlocking), free_unique_id, status,
+                   tl_team->context->lib);
+    nccl_status = ncclCommInitRank(&team->nccl_comm,team->size,
+                                   team->unique_id[0], team->rank);
+    if (nccl_status != ncclSuccess) {
+        tl_info(tl_team->context->lib, "NCCL error %d %s",
+                nccl_status, ncclGetErrorString(nccl_status));
+        status = UCC_ERR_NO_MESSAGE;
+        goto free_stream;
+    }
+    ucc_free(team->unique_id);
+    tl_info(tl_team->context->lib, "initialized tl team: %p", team);
+    return UCC_OK;
+
+free_stream:
+    cudaStreamDestroy(team->stream);
+free_unique_id:
+    ucc_free(team->unique_id);
+    return status;
+}
+
+static ucc_status_t ucc_tl_nccl_coll_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task  = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_status_t       status = UCC_OK ;
+
+    tl_info(UCC_TL_TEAM_LIB(task->team), "finalizing coll task %p", task);
+    CUDACHECK_GOTO(cudaEventDestroy(task->completed), put_task, status,
+                   UCC_TL_TEAM_LIB(task->team));
+put_task:
+    ucc_mpool_put(task);
+    return status;
+}
+
+ucc_status_t ucc_tl_nccl_coll_init(ucc_base_coll_args_t *coll_args,
+                                   ucc_base_team_t *team,
+                                   ucc_coll_task_t **task_h)
+{
+    ucc_tl_nccl_team_t    *nccl_team = ucc_derived_of(team, ucc_tl_nccl_team_t);
+    ucc_tl_nccl_context_t *nccl_ctx  = ucc_derived_of(team->context,
+                                                      ucc_tl_nccl_context_t);
+    ucc_tl_nccl_task_t    *task;
+    ucc_status_t status;
+
+    task = ucc_mpool_get(&nccl_ctx->req_mp);
+    ucc_coll_task_init(&task->super);
+    memcpy(&task->args, &coll_args->args, sizeof(ucc_coll_args_t));
+    task->team = nccl_team;
+    task->super.finalize = ucc_tl_nccl_coll_finalize;
+    CUDACHECK_GOTO(cudaEventCreateWithFlags(&task->completed,
```

Comment:
is cudaEventCreateWithFlags costly? Do you want another mpool for that? (i think Devendar used a pool for cuda events in hcoll's nccl bcol, but i don't know if it is needed)